pennylane_lightning/experimental/_adjoint_jacobian.py

```diff
@@ -0,0 +1,204 @@
+
+
+def _process_jacobian_tape(tape, state):
+    """Process a Jacobian tape before calculation.
+
+    Args:
+        tape (QuantumTape): A quantum tape recording a variational quantum program.
+        state(np.array): unravelled initial state (1D)
+
+    Returns:
+        Set: A set with a tape processed for Lightning.
+    """
+    use_csingle = True if state.dtype == np.complex64 else False
+    # To support np.complex64 based on the type of self._state
+    create_ops_list = (
+        adjoint_diff.create_ops_list_C64 if use_csingle else adjoint_diff.create_ops_list_C128
+    )
+
+    obs_serialized = _serialize_observables(tape, use_csingle=use_csingle)
+    ops_serialized, use_sp = _serialize_ops(tape)
+
+    ops_serialized = create_ops_list(*ops_serialized)
+
+    # We need to filter out indices in trainable_params which do not
+    # correspond to operators.
+    trainable_params = sorted(tape.trainable_params)
+    if len(trainable_params) == 0:
+        return None
+
+    tp_shift = []
+    record_tp_rows = []
+    all_params = 0
+
+    for op_idx, tp in enumerate(trainable_params):
+        op, _ = tape.get_operation(
+            op_idx, False
+        )  # get op_idx-th operator among differentiable operators
+        if isinstance(op, Operation) and not isinstance(op, (BasisState, QubitStateVector)):
+            # We now just ignore non-op or state preps
+            tp_shift.append(tp)
+            record_tp_rows.append(all_params)
+        all_params += 1
+
+    if use_sp:
+        # When the first element of the tape is state preparation. Still, I am not sure
+        # whether there must be only one state preparation...
+        tp_shift = [i - 1 for i in tp_shift]
+
+    state_vector = StateVectorC64(state) if use_csingle else StateVectorC128(state)
+    return {
+        "state_vector": state_vector,
+        "obs_serialized": obs_serialized,
+        "ops_serialized": ops_serialized,
+        "tp_shift": tp_shift,
+        "record_tp_rows": record_tp_rows,
+        "all_params": all_params,
+    }
+
+
+def adjoint_jacobian(tape, c_dtype=np.complex128, starting_state=None):
+    """Calculates the Adjoint Jacobian for a given tape.
+
+    Args:
+        tape (QuantumTape): A quantum tape recording a variational quantum program.
+        c_dtype (Complex data type, Optional): Default to ``np.complex128``.
+        starting_state (np.array, Optional): unravelled initial state (1D). Default to None.
+
+    Returns:
+        np.array: An array results.
+    """
+    if starting_state:
+        state = starting_state
+    else:
+        state = create_initial_state(tape.wires, tape._prep[0] if tape._prep else None)
+        state = np.ravel(asarray(state, c_dtype))
+        state = apply_operations(tape._ops, state)
+
+    if len(tape.measurements) == 0:  # the tape does not have measurements
+        return np.array([], dtype=state.dtype)
+
+    _check_adjoint_method_supported(tape)
```

Comment:
We can now move these checks into preprocessing so they happen earlier in the pipeline. We want to fail fast, and not do more work than we have to if we are just going to fail anyway.

Especially as this is run *after* the forward pass execution, we don't want to perform the forward pass execution only to fail and lose that information.

Reply:
I will do that. This is the main reason I refactored this part of the code, but I wanted some feedback before implementing it.
Thank you.